src/gsi/setupq.f90

```diff
@@ -549,10 +579,36 @@ subroutine setupq(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsav
      rmaxerr=max(small2,rmaxerr)
      errorx =(data(ier,i)+dprpx)*qsges
 
-! Interpolate guess moisture to observation location and time
-     call tintrp31(ges_q,qges,dlat,dlon,dpres,dtime, &
-        hrdifsig,mype,nfldsig)
-    
+! qges: Interpolate guess moisture to observation location and time 
+
+     if (.not. ( hofx_2m_sfcfile .and. landsfctype) ) then 
+        call tintrp31(ges_q,qges,dlat,dlon,dpres,dtime, &
+           hrdifsig,mype,nfldsig)
+     else
+        ! only use land locations
+        if (int(data(idomsfc,i)) .NE. 1  ) muse(i) = .false.
+
+        call tintrp2a11(ges_q2m,qges,dlat,dlon,dtime,hrdifsig,mype,nfldsig)
+
+        ! terrain correction: assume RH_zo = RH_zm, and correct T with 
+        ! same lapse rate as used for T2m terrain correction
+
+        delta_z = data(istnelv,i) - data(izz,i) ! obs -model
+
+        do jj=1,nfldsig
+           ! qsat in model at height of obs
+           call genqsat(qg2m_o(:,:,jj),ges_t2m(:,:,jj)+delta_z*T_lapse,ges_ps(:,:,jj),lat2,lon2,1,ice,iderivative)
+        enddo
+
+        call tintrp2a11(qg2m_o,qsges_o,dlat,dlon,dtime,hrdifsig,mype,nfldsig)
+        q_delta_terrain = (qsges/qsges_o - 1)*qob
+        qob = qob * ( qsges/qsges_o)
+
+        !update the station elevation
+        data(istnelv,i) = data(izz,i)
+
+     endif
+
```

Comment:
I have a concern on setupq.f90:
It is straightforward to apply terrain adjustment for 2mT based on the lapse rate fact. However, it is not the same situation for 2mQ. When one brings an airmass from height1 to height2, the Q will not change (in spite of qsat changes) unless supersaturation happens. It is beneficial to check and address supersaturation after the minimization, but I doubt we need to directly change the original 2mQ observations.